hivemind/utils/mpfuture.py

```diff
@@ -2,171 +2,253 @@
-                self._state, self._exception = base.FINISHED, e
-
-    def _await_terminal_state(self, timeout: Optional[float]):
-        """ Await updates until future is either finished, cancelled or got an exception """
-        time_left = float('inf') if timeout is None else timeout
-        time_before = time.monotonic()
-        while self._state not in self.TERMINAL_STATES and time_left > 0:
-            self._recv_updates(time_left if timeout else None)
-            time_spent = time.monotonic() - time_before
-            time_left, time_before = time_left - time_spent, time_before + time_spent
-
-    def _sync_updates(self):
-        """ Apply queued updates from a paired MPFuture without waiting for new ones """
+    @property
+    def _state(self) -> State:
+        shared_state = ALL_STATES[self._shared_state_code.item()]
+        return self._state_cache.get(shared_state, shared_state)
+
+    @_state.setter
+    def _state(self, new_state: State):
+        self._shared_state_code[...] = ALL_STATES.index(new_state)
+        if self._state in TERMINAL_STATES and self._loop is not None and not self._aio_event.is_set():
+            self._set_event_threadsafe()
+
+    def _set_event_threadsafe(self):
         try:
-            self._recv_updates(timeout=0)
-        except TimeoutError:
-            pass
+            loop = asyncio.get_running_loop()
+        except RuntimeError:
+            loop = None
+        if loop == self.get_loop():
+            asyncio.create_task(self._event_setter())
+        else:
+            asyncio.run_coroutine_threadsafe(self._event_setter(), self._loop)
+
+    async def _event_setter(self):
+        self._aio_event.set()
+
+    @property
+    def _sender_pipe(self) -> mp.connection.Connection:
+        """ a pipe that can be used to send updates to the MPFuture creator """
+        return self.global_mpfuture_receivers[self._origin_pid]
+
+    @classmethod
+    def _initialize_mpfuture_backend(cls):
+        pid = os.getpid()
+        logger.debug(f"Initializing MPFuture backend for pid {pid}")
+        assert pid != cls.active_pid and pid not in cls.global_mpfuture_receivers, "already initialized"
+
+        recv_pipe, send_pipe = mp.Pipe(duplex=False)
+        cls.active_pid, cls.active_futures, cls.global_mpfuture_receivers[pid] = pid, {}, send_pipe
+        cls.pipe_waiter_thread = threading.Thread(target=cls._process_updates_in_background, args=[recv_pipe],
+                                                  name=f'{__name__}.BACKEND', daemon=True)
+        cls.pipe_waiter_thread.start()
+
+    @classmethod
+    def _process_updates_in_background(cls, receiver_pipe: mp.connection.Connection):
+        pid = os.getpid()
+        while True:
+            try:
+                uid, update_type, payload = receiver_pipe.recv()
+                if uid not in cls.active_futures:
+                    logger.debug(f"Ignoring update to future with uid={uid}: the future is no longer active.")
```

Comment:
This message uses the term "active future" that is not standard and not clear for a user. AFAIU, not active = already done or destroyed, so I suggest to rewrite it as:

```suggestion
                    logger.debug(f"Ignoring update to future with uid={uid}: the future is already done or destroyed.")
```

Reply:
If there is an easy way to distinguish the "already done" and "destroyed" states here, I'd suggest to do it and log more specific message (should be useful for debug).

Also, what do you think of making this a `logger.warning`? (asyncio's non-awaited tasks cause a warning in this case)